x/snapshot/keeper/keeper.go

```diff
@@ -278,51 +289,39 @@ func (k Keeper) executeSnapshot(ctx sdk.Context, counter int64, keyRequirement t
 		)
 	}
 
-	corruptionThreshold := tss.ComputeAbsCorruptionThreshold(keyRequirement.SafetyThreshold, totalShareCount)
-	if corruptionThreshold < 0 ||
-		corruptionThreshold >= totalShareCount.Int64() {
-		return exported.Snapshot{}, fmt.Errorf("invalid corruption threshold: %d, total share count: %d", corruptionThreshold, totalShareCount.Int64())
+	snapshot := exported.NewSnapshot(
+		participants,
+		ctx.BlockTime(),
+		ctx.BlockHeight(),
+		totalShareCount,
+		counter,
+		keyRequirement.KeyShareDistributionPolicy,
+		tss.ComputeAbsCorruptionThreshold(keyRequirement.SafetyThreshold, totalShareCount),
+	)
+	if err := snapshot.Validate(); err != nil {
+		return exported.Snapshot{}, err
 	}
 
-	snapshot := exported.Snapshot{
-		Validators:                 participants,
-		Timestamp:                  ctx.BlockTime(),
-		Height:                     ctx.BlockHeight(),
-		TotalShareCount:            totalShareCount,
-		Counter:                    counter,
-		KeyShareDistributionPolicy: keyRequirement.KeyShareDistributionPolicy,
-		CorruptionThreshold:        corruptionThreshold,
-	}
-	ctx.KVStore(k.storeKey).Set(counterKey(counter), k.cdc.MustMarshalLengthPrefixed(&snapshot))
+	k.setSnapshot(ctx, snapshot)
 
 	return snapshot, nil
 }
 
-func (k Keeper) setLatestCounter(ctx sdk.Context, counter int64) {
-	bz := make([]byte, 8)
-	binary.LittleEndian.PutUint64(bz, uint64(counter))
-
-	ctx.KVStore(k.storeKey).Set([]byte(lastCounterKey), bz)
-}
-
-func counterKey(counter int64) []byte {
-	return []byte(fmt.Sprintf("%s%d", counterPrefix, counter))
-}
-
 // RegisterProxy registers a proxy address for a given operator, which can broadcast messages in the principal's name
 // The proxy will be marked as active and to be included in the next snapshot by default
 func (k Keeper) RegisterProxy(ctx sdk.Context, operator sdk.ValAddress, proxy sdk.AccAddress) error {
 	if bytes.Equal(operator, proxy) {
 		return fmt.Errorf("proxy address cannot be the same as the operator address")
 	}
 
-	if storedOperator := k.GetOperator(ctx, proxy); !storedOperator.Empty() && !bytes.Equal(operator, storedOperator) {
-		return fmt.Errorf("address %s already registered as a proxy to another operator", proxy.String())
-	}
-
-	if storedProxy, _ := k.GetProxy(ctx, operator); !storedProxy.Empty() && !bytes.Equal(storedProxy, proxy) {
-		return fmt.Errorf("proxy mismatch (operator %s registered proxy %s, received %s)",
-			operator.String(), sdk.AccAddress(storedProxy).String(), proxy.String())
+	if validatorProxy, ok := k.getValidatorProxy(ctx, operator); ok && (!validatorProxy.Validator.Equals(operator) || !validatorProxy.Proxy.Equals(proxy)) {
```

Comment:
this is very succinct but a bit confusing to read to me. What do you think about something like this?
```
fromOperator, operatorExists := k.getValidatorProxy(ctx, operator)
newVal := types.NewValidatorProxy(operator, proxy, true)
if operatorExists && !fromOperator.Equals(newVal){
...
}
```

Reply:
Let me give it a try.